browser/ui/webui/brave_vpn/vpn_panel_ui.cc

```diff
@@ -0,0 +1,83 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/brave_vpn/vpn_panel_ui.h"
+
+#include <memory>
+#include <utility>
+
+#include "base/bind.h"
+#include "brave/browser/brave_vpn/brave_vpn_service_factory.h"
+#include "brave/common/webui_url_constants.h"
+#include "brave/components/brave_vpn/resources/panel/grit/brave_vpn_panel_generated_map.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "chrome/browser/ui/webui/favicon_source.h"
+#include "chrome/browser/ui/webui/webui_util.h"
+#include "components/favicon_base/favicon_url_parser.h"
+#include "components/grit/brave_components_resources.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_data_source.h"
+namespace {
+
+content::WebContents* GetWebContentsFromTabId(int32_t tab_id) {
+  for (auto* browser : *BrowserList::GetInstance()) {
+    TabStripModel* tab_strip_model = browser->tab_strip_model();
+    for (int index = 0; index < tab_strip_model->count(); ++index) {
+      content::WebContents* contents = tab_strip_model->GetWebContentsAt(index);
+      if (sessions::SessionTabHelper::IdForTab(contents).id() == tab_id) {
+        return contents;
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+}  // namespace
+
+VPNPanelUI::VPNPanelUI(content::WebUI* web_ui)
+    : ui::MojoBubbleWebUIController(web_ui, true) {
+  content::WebUIDataSource* source =
+      content::WebUIDataSource::Create(kVPNPanelHost);
+
+  webui::SetupWebUIDataSource(
+      source,
+      base::make_span(kBraveVpnPanelGenerated, kBraveVpnPanelGeneratedSize),
+      IDR_VPN_PANEL_HTML);
+  content::WebUIDataSource::Add(web_ui->GetWebContents()->GetBrowserContext(),
+                                source);
+  Profile* profile = Profile::FromWebUI(web_ui);
+  content::URLDataSource::Add(
+      profile, std::make_unique<FaviconSource>(
+                   profile, chrome::FaviconUrlFormat::kFavicon2));
+}
+
+VPNPanelUI::~VPNPanelUI() = default;
+
+WEB_UI_CONTROLLER_TYPE_IMPL(VPNPanelUI)
+
+void VPNPanelUI::BindInterface(
+    mojo::PendingReceiver<brave_vpn::mojom::PanelHandlerFactory> receiver) {
+  panel_factory_receiver_.reset();
+  panel_factory_receiver_.Bind(std::move(receiver));
+}
+
+void VPNPanelUI::CreatePanelHandler(
+    mojo::PendingRemote<brave_vpn::mojom::Page> page,
+    mojo::PendingReceiver<brave_vpn::mojom::PanelHandler> panel_receiver) {
+  DCHECK(page);
+  auto* profile = Profile::FromWebUI(web_ui());
+  DCHECK(profile);
+
+  panel_handler_ = std::make_unique<VPNPanelHandler>(
+      std::move(panel_receiver), this,
+      base::BindRepeating(&GetWebContentsFromTabId),
```

Comment:
Just curious why this is passed as a callback.
I think `VPNPanelHandler` can implement it instead of getting as an args.
Also, it seems `VPNPanelHandler` doesn't use this callback.

Reply:
if you're referring to the `GetWebContentsFromTabId` then that makes sense. I've removed that. however, the panel receiver is needed and without it the panel handler cannot seem to open/close the panel.